### PR TITLE
i#5694 core-sharded: Add core-serial support

### PR DESCRIPTION
**Pull request by derekbruening**
Adds a new scheduler option single_lockstep_output which multiplexes the virtual core output streams onto a single global stream.  This is simple to implement as the existing scheduler_t::stream_t class already multiplexes inputs onto an output.

Hooks up the drcachesim launcher -core_serial option to this new scheduler mode.

Updates the schedule_stats, basic_counts, and cache_simulator tools to support core_serial.  For cache_simulator, the existing thread-to-core mapping code for round-robin and for -cpu_scheduling is kept for when in thread-sharded mode; in core-sharded mode, the scheduler's cpuid is mapped to a core index.

Adds a core_serial test of schedule_stats and basic_counts and a test of cache_simulator using the scheduler's -cpu_schedule_file as-traced mode.

Adds some dr$sim unit tests for cpuid to core mapping and error modes.

Issue: #5694

### CODE REVIEW

**Comment by derekbruening** [comment]
win64 failure is replaceall #5412

**Review by abhinav92003 — APPROVED** [review]

File: clients/drcachesim/analyzer.cpp [hunk 1]
```diff
@@ -291,15 +292,23 @@ analyzer_tmpl_t<RecordType, ReaderType>::init_scheduler_common(
         sched_ops = std::move(options);
         if (sched_ops.quantum_unit == sched_type_t::QUANTUM_TIME)
             sched_by_time_ = true;
+        if (!parallel_) {
+            // output_count remains the # of virtual cores, but we have just
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/analyzer.cpp:296]: Yes, there is a comment on line 287 about this.  Looks like it wasn't in the header: added there.

File: clients/drcachesim/scheduler/scheduler.h [hunk 2]
```diff
@@ -591,6 +591,14 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
         std::unique_ptr<ReaderType> kernel_switch_reader;
         /** The end reader for #kernel_switch_reader. */
         std::unique_ptr<ReaderType> kernel_switch_reader_end;
+        /**
+         * If true, enables a mode where all outputs are serialized into one global outer
+         * layer output.  The single global output stream alternates in round-robin
+         * lockstep amoung each core output.  The inner outputs operate just like they
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:597]: Done.

[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:597]: Reworded "inner" to "core" and "other" to "other core" so it's clear they're all the same.

**Pull request merged**